SimPEG/electromagnetics/static/utils/static_utils.py

```diff
@@ -687,29 +695,345 @@ def plot_3d_pseudosection(
+    locations_n : numpy.array
+        An (n, dim) numpy array containing N electrode locations. If None,
+        we assume all receivers are Pole receivers.
+    data_type : str
+        Must be one of {'volt', 'apparent_conductivity', 'apparent_resistivity', 'apparent_chargeability'}
+    output_sorting : bool
+        This option is used if the ABMN locations are sorted during the creation of the survey
+        and you would like to sort any data vectors associated with the electrode locations.
+        If False, the function will output a SimPEG.electromagnetic.static.survey.Survey object.
+        If True, the function will output a tuple containing the survey object and a numpy array
+        (n,) that will sort the data vector to match the order of the electrodes in the survey.
+
+
+    Returns
+    -------
+    survey
+        A SimPEG.electromagnetic.static.survey.Survey object
+    sort_index
+        A numpy array which defines any sorting that took place when creating the survey
+
+
+    """
+    
+    if locations_a is None:
+        raise AssertionError("Locations for A electrodes must be provided.")
+    if locations_m is None:
+        raise AssertionError("Locations for M electrodes must be provided.")
+
+    assert data_type.lower() in [
+        "volt",
+        "apparent_conductivity",
+        "apparent_resistivity",
+        "apparent_chargeability",
+    ], "data_type must be one of 'volt', 'apparent_conductivity', 'apparent_resistivity', 'apparent_chargeability'"
+
+    if locations_b is None:
+        locations_b = locations_a
+
+    if locations_n is None:
+        locations_n = locations_m
+    
+    if (locations_a.shape==locations_b.shape==locations_m.shape==locations_n.shape) == False:
+        raise AssertionError("Arrays containing A, B, M and N electrode locations must be same shape.")
+
+    # Set up keeping track of sorting of rows and unique sources
+    n_rows = np.shape(locations_a)[0]
+    k = np.arange(0, n_rows)
+    out_indices = []
+    unique_ab, ab_index = np.unique(np.c_[locations_a, locations_b], axis=0, return_index=True)
+    ab_index = np.sort(ab_index)
+    
+    # Loop over all unique source locations
+    source_list = []
+    for ii, ind in enumerate(ab_index):
+            
+        # Get source location
+        src_loc_a = mkvc(locations_a[ind, :])
+        src_loc_b = mkvc(locations_b[ind, :])
+        
+        # Get receiver locations
+        rx_index = np.where(
+            (
+                (np.sqrt(np.sum((locations_a - src_loc_a)**2, axis=1)) < 1e-3) &
+                (np.sqrt(np.sum((locations_b - src_loc_b)**2, axis=1)) < 1e-3)
+            )
+        )[0]
+        
+        rx_loc_m = locations_m[rx_index, :]
+        rx_loc_n = locations_n[rx_index, :]
+        
+        # Extract pole and dipole receivers
+        k_ii = k[rx_index]
+        is_pole_rx = np.all(np.isclose(rx_loc_m, rx_loc_n, atol=1e-3), axis=1)
```

Comment:
Are there any other conditions we would want to accept as indicative of a `pole` source/receiver? Maybe if B or N is infinity or NaN?

Reply:
I am happy the way it is. But I think once this is released, the community may offer feedback which could lead to some small additions.